test/constrained-generics/hashtable/MyHashtable.chpl

```diff
@@ -315,13 +324,119 @@ module ChapelHashtable {
       // Free the buffer
       _freeData(table, tableSize);
     }
+  }  // record chpl__hashtable
+
+
+/////////////////////////////////////////////////////////////////////////////
+// interface definitions and implements statements
+
+//
+// Hashable: user-facing requirements on the hashtable key type.
+//
+interface Hashable(Key) {
+  //  proc hash(arg: Key): uint;
+  proc chpl__defaultHashWrapper(arg: Key): int;
+  operator ==(lhs: Key, rhs: Key): bool;
+}
+
+//
+// StdOps: various common operations we expect.
+//
+// Todo: include == and resolve the ambiguity with Hashable.==
+// possibly by requiring Hashable to implement StdOps.
+//
+interface StdOps(Val) {
+  proc chpl__initCopy(arg: Val, definedConst: bool): Val;
+  operator =(ref lhs: Val, rhs: Val);
+  proc toString(arg: Val): string; // so we can write it out
+}
+
+// These use built-in implementations and proc toString below.
+string implements Hashable;
+string implements StdOps;
+int implements Hashable;
+int implements StdOps;
+
+// The default implementation for toString().
+proc toString(arg): string return arg: string;
+
+//
+// chpl_Hashtable: intended to replace the uses of the type chpl__hashtable.
+//
+// It contains both public functionality and implementation details.
+//
+interface chpl_Hashtable(HT) {
+  type keyType;
+  type valType;
+  keyType implements Hashable;
+  keyType implements StdOps;
+  valType implements StdOps;
+  // can't have == in StdOps because then == on keys would be ambiguous
```

Comment:
This ambiguity issue is interesting. Could you explain it more (here in the PR discussion, not in source code comments) ?